### PR TITLE
tests: Add putObject and makeBucket negative tests

### LINKED ISSUE #2261: mc cp functional test fails on gateways with latest build
## Expected behaviour
mc cp tests should pass.
## Actual behaviour
failure is logged without indication of what failed - all 3 gateways fail for this test. 
error.log
s3 gateway along shows this in error log 
md5sum: /tmp/datafile-1-MB: No such file or directory

output.log
{"name": "mc", "duration": "2623", "function": "./mc cp /mint/data/datafile-1-MB target/mc-mint-test-bucket-31520//2123123\123", "status": "FAIL", "error": "`/mint/data/datafile-1-MB` -> `target/mc-mint-test-bucket-31520/2123123\123`
Total: 1024.00 KB, Transferred: 1024.00 KB, Speed: 911.99 KB/s"}
## mc version
- (paste output of `mc version`)

## System information
Mint image built from  commit 158108152fc86c52d5543995e0e7634bf1b6da74
Minio version:
Version: 2017-09-25T21:57:17Z
Release-Tag: RELEASE.2017-09-25T21-57-17Z
Commit-ID: d3eb5815d9951fdb82d0134b7651368bab71d72e

### PR DESCRIPTION
**Pull request by vadmeste**
Add new negative tests to functional tests. For that purpose, two functions are
introduced. assert_success makes sure that the passed command returns 0 exit code
and raise an error otherwise. assert_failure makes sure that the passed
command fails and raises an error otherwise.

Fixes #2261

### CODE REVIEW

File: functional-tests.sh [hunk 1]
```diff
@@ -202,22 +212,33 @@ function test_make_bucket()
...
+
+function test_make_bucket_error() {
+    show "${FUNCNAME[0]}"
+
+    start_time=$(get_time)
+    bucket_name="Mc-test-bucket-$RANDOM"
+    assert_failure "$start_time" "${FUNCNAME[0]}" mc_cmd mb "${SERVER_ALIAS}/${bucket_name}"
 
     log_success "$start_time" "${FUNCNAME[0]}"
+
```
Comment (balamurugana) [inline functional-tests.sh]: You could remove unnecessary new lines.

File: functional-tests.sh [hunk 2]
```diff
@@ -202,22 +212,33 @@ function test_make_bucket()
...
+    assert_success "$start_time" "${FUNCNAME[0]}" mc_cmd rm "${SERVER_ALIAS}/${bucket_name}"
+
+    log_success "$start_time" "${FUNCNAME[0]}"
+}
+
+function test_make_bucket_error() {
+    show "${FUNCNAME[0]}"
+
+    start_time=$(get_time)
+    bucket_name="Mc-test-bucket-$RANDOM"
```
Comment (balamurugana) [inline functional-tests.sh]: Could you add more invalidation than single `M`?

File: functional-tests.sh [hunk 3]
```diff
@@ -226,8 +247,21 @@ function test_put_object()
...
+}
+
+function test_put_object_error()
+{
+
+    show "${FUNCNAME[0]}"
+    start_time=$(get_time)
+
+    long_name=$(printf '=%.0s' {1..1100})
+    object_name="mc-test-object-$long_name"
```
Comment (balamurugana) [inline functional-tests.sh]: I think `object_name=$(printf "mc-test-object-%01100d" 1)` would be simpler and understandable.

File: functional-tests.sh [hunk 4]
```diff
@@ -202,22 +212,33 @@ function test_make_bucket()
 
     start_time=$(get_time)
     bucket_name="mc-test-bucket-$RANDOM"
-    assert_log_failure "$start_time" "${FUNCNAME[0]}" mc_cmd mb "${SERVER_ALIAS}/${bucket_name}"
-    assert_log_failure "$start_time" "${FUNCNAME[0]}" mc_cmd rm "${SERVER_ALIAS}/${bucket_name}"
+    assert_success "$start_time" "${FUNCNAME[0]}" mc_cmd mb "${SERVER_ALIAS}/${bucket_name}"
```
[the remark this answers is not in the document]
Comment (vadmeste) [reply functional-tests.sh]: > Current function assert_log_failure means assert on failure with log. assert_success gives completely different meaning

AFAIK, assert means confirm, assert_success means confirm command success and assert_failure means confirm command failure

File: functional-tests.sh [hunk 5]
```diff
@@ -135,16 +135,18 @@ function fail()
     return "$rv"
 }
 
-function assert_log_failure()
+function assert()
```
[the remark this answers is not in the document]
Comment (vadmeste) [reply functional-tests.sh]: > You could name assert_log indicating this function logs and asserts.

Most assert() that I saw print log. I think this is better than assert_log() which can lead to more confusion.

[the remark this answers is not in the document]
Comment (balamurugana) [reply functional-tests.sh]: OK. Can you send a separate PR for the name fix?

[the remark this answers is not in the document]
Comment (vadmeste) [reply functional-tests.sh]: Besides, assert_log() doesn't always log.

File: functional-tests.sh (hunk 2 above)
Comment (vadmeste) [reply functional-tests.sh]: > Could you add more invalidation than single M?

Done

**Comment by vadmeste** [comment]
All comments addressed @balamurugana

File: functional-tests.sh [hunk 6]
```diff
@@ -208,6 +218,19 @@ function test_make_bucket()
...
 }
 
+function test_make_bucket_error() {
+    show "${FUNCNAME[0]}"
+
+    start_time=$(get_time)
+    bucket_name="Mc-test-bucket-$RANDOM"
+    assert_log_success "$start_time" "${FUNCNAME[0]}" mc_cmd mb "${SERVER_ALIAS}/${bucket_name}"
+
+    bucket_name="mc%test%bucket%$RANDOM"
```
Comment (balamurugana) [inline functional-tests.sh]: I was expecting one `mb` with invalid bucket name something like `"MC-test%bucket%$RANDOM"`

Comment (vadmeste) [reply functional-tests.sh]: Done

**Pull request merged**